feat: Add Replay API support

nieomylnieja opened the pull request:
## Motivation

Replay clients currently maintain separate HTTP requests and models. A versioned SDK interface provides one public contract for managing Replays and reduces drift between consumers.

## Summary

- Added Replay V1 operations for run, cancel, delete, list, status, and availability.
- Added typed request, response, status, cancellation, source, duration, and availability values with request validation.
- Preserved RFC3339 `createdAt` in Replay list items and retained unknown availability reasons as raw values.
- Removed the legacy Replay models from `sdk/models` and removed the SDK-side maximum Replay duration limit.

## Related changes

- [sloctl #482](https://github.com/nobl9/sloctl/pull/482)
- [n9 #21542](https://github.com/nobl9/n9/pull/21542)

## Testing

- `make test` and `make check` pass.
- Focused Replay e2e passed on `dev-demo-1`, covering SLO- and data-source-based availability, queued and in-progress states, RFC3339 `createdAt`, queued deletion, status retrieval, source-SLO mappings, and cancellation behavior.
- Two focused Replay e2e runs passed concurrently to exercise queue contention and unrelated scheduler wakeups.
- In the full e2e suite, Replay passed; five unrelated environment-dependent tests failed in existing object, annotation, SLO release-channel, and Prometheus fixture flows.

## Release Notes

Added a versioned Replay API for running, listing, canceling, deleting, and inspecting Replays. Replay models now live under `sdk/endpoints/replay/v1`.

## Breaking Changes

Removed Replay models from `sdk/models`. Consumers must import Replay request and response types from `sdk/endpoints/replay/v1`.

nieomylnieja committed:
add replay endpoints

nieomylnieja committed:
Merge remote-tracking branch 'origin/main' into add-replay-endpoints

nieomylnieja committed:
feat: refactor replay endpoints and validation logic

This commit introduces significant changes to the replay endpoints and their associated validation logic. The following updates have been made:

1. **Endpoint Updates**:
   - Added new endpoints for `Cancel`, `Delete`, `DeleteAll`, `List`, and `GetStatus` operations.
   - Refactored the `Replay` endpoint to `Run` for better clarity and alignment with its functionality.
   - Introduced a helper function `handleReplayError` to centralize error handling logic for replay-related operations.

2. **Request and Validation Changes**:
   - Renamed `ReplayRequest` to `RunRequest` and updated its fields for consistency (`Slo` to `SLO`, `ReplayDuration` to `Duration`, etc.).
   - Added new request types: `DeleteRequest`, `CancelRequest`, and `GetStatusRequest`.
   - Updated validation rules to reflect the new structure and naming conventions.
   - Enhanced validation logic to ensure only one of `duration` or `startDate` is set, and added checks for future dates.

3. **Response Updates**:
   - Updated `ReplayWithStatus` to use `SLO` instead of `Slo`.
   - Introduced a new response type `ReplayListItem` for listing replay items.

4. **Test Updates**:
   - Refactored test cases to align with the new request and response structures.
   - Updated test names and assertions to reflect the renaming and added validation rules.

5. **Miscellaneous**:
   - Updated helper functions and constants to align with the new naming conventions.
   - Improved error messages and validation error codes for better clarity.

nieomylnieja committed:
Merge remote-tracking branch 'origin/main' into add-replay-endpoints

# Conflicts:
#	sdk/endpoints/replay/v1/request_test.go
#	sdk/models/replay.go

nieomylnieja committed:
Merge branch 'main' into add-replay-endpoints

nieomylnieja committed:
refactor: move replay and prometheus APIs into client

Consolidates versioned API accessors on `sdk/client.go` and removes the old duplicate entry points.
Drops deprecated replay models from `sdk/models` in favor of the endpoint-specific types.

nieomylnieja committed:
refactor: simplify replay v1 package declarations

Update the package comment to match the actual package name and remove
redundant generic type arguments from replay request validators.

nieomylnieja committed:
refactor: simplify replay delete requests

Remove the dedicated `DeleteAll` endpoint and fold full-delete behavior into `DeleteRequest`.
Also align replay request payloads with explicit JSON fields and drop unsupported `EndDate`.

ditrytus commented on sdk/endpoints/replay/v1/response.go:
```diff
@@ -0,0 +1,67 @@
+	RetrievedScope string `json:"retrievedScope,omitempty"`
+	RetrievedFrom  string `json:"retrievedFrom,omitempty"`
+	Status         string `json:"status"`
+	Cancellation   string `json:"cancellation,omitempty"`
```
This field also has a predefined list of viariants.

nieomylnieja replied:
@ditrytus Added a `ReplayListStatus...` constant group for `ReplayListItem.Status`, mirroring the known `failed` and `completed` status variants without changing the response field type.

nieomylnieja replied:
@ditrytus Follow-up: changed `ReplayStatus.Status`, `ReplayStatus.Cancellation`, `ReplayListItem.Status`, and `ReplayListItem.Cancellation` to named string enum types. Added typed constants for the known status variants and the known cancellation variant (`blocked`).

nieomylnieja replied:
@ditrytus Follow-up after checking the n9 implementation: `source` now uses `ReplaySource` (`user`, `error_budget_adjustment`), `replayType`/availability `type` use `ReplayType` (`reimport_and_recalculation`, `recalculation`), `ReplayListItem.Status` uses the list status set (`unknown`, `queued`, `in progress`, `completed`, `failed`, `canceled`), and `ReplayStatus.Cancellation`/list cancellation use the full cancellation set (`possible`, `blocked`, `requested`, `denied`, `done`). I also added the fixed availability reasons currently defined in n9.

ditrytus replied:
Cancellation is fine, but status needs to be changed. I'll describe it in a separate comment: https://github.com/nobl9/nobl9-go/pull/756#discussion_r3497820299

nieomylnieja replied:
@ditrytus Follow-up handled in the status thread: `ReplayProcessStatus` is now trimmed to the coarse n9 public set and `ReplayListStatus` aliases it. Cancellation remains aligned with n9: `possible`, `blocked`, `requested`, `denied`, and `done`.

ditrytus commented on sdk/endpoints/replay/v1/endpoints.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+func (e endpoints) Delete(ctx context.Context, params DeleteRequest) (err error) {
+	return e.deleteReplay(ctx, params)
```
`deleteReplay` can be inlined here. It is not used anywhere else and `Delete` is now just a passthrough function.

nieomylnieja replied:
@ditrytus Inlined the delete request implementation into `Delete` and removed the unused `deleteReplay` helper.

ditrytus commented on sdk/endpoints/replay/v1/request.go:
```diff
@@ -0,0 +1,244 @@
+	SLO     string `json:"slo"`
+}
+
+type GetAvailabilityRequest struct {
+	Project           string
+	DataSourceProject string
+	DataSource        string
+	DataSourceKind    string
+	SLOName           string
+	Type              string
+	DurationUnit      string
+	DurationValue     int
+}
```
Doesn't this struct don't require struct tags? Maybe if this one doesn't the others don't as well?

ditrytus replied:
What about this question?

nieomylnieja replied:
this one is only used as a intermediate input object for the users, it is not used as the actual payload, that's why it doesn't have the tags

ditrytus reviewed (changes requested):
There is a regression in relation to this change: https://github.com/nobl9/nobl9-go/pull/845 (30 day limit enforced in SDK).

nieomylnieja committed:
feat: remove replay duration limit and add replay e2e coverage

Allow replay requests longer than 30 days by dropping the SDK-side cap on duration and start time validation.
Replace endpoint stub tests with an end-to-end replay test that exercises availability, run, list, status, and delete flows.

nieomylnieja committed:
Merge branch 'main' into add-replay-endpoints

nieomylnieja committed:
Merge branch 'main' into add-replay-endpoints

nieomylnieja committed:
feat: add typed replay statuses

Expose replay process and cancellation states as typed constants in replay
responses and inline the delete endpoint implementation without changing behavior

nieomylnieja committed:
Merge remote-tracking branch 'origin/main' into add-replay-endpoints

nieomylnieja committed:
feat: add typed replay request and response enums

Validate replay source and replay type in run requests and
model replay availability and list statuses with dedicated types.
Update replay tests to cover the new request fields and response parsing.

nieomylnieja committed:
fix: align replay status types across responses

Collapse replay process and list statuses to the same coarse public values.
Add coverage for availability query encoding and replay status unmarshalling.

nieomylnieja committed:
chore: add exception to cspell

ditrytus commented on sdk/endpoints/replay/v1/response.go:
```diff
@@ -9,21 +9,21 @@ type ReplayWithStatus struct {
+	Cancellation ReplayCancellationStatus `json:"cancellation"`
+	CanceledBy   string                   `json:"canceledBy,omitempty"`
+	TriggeredBy  string                   `json:"triggeredBy"`
+	Unit         string                   `json:"unit"`
```
This type can be changed from `string` to `DurationUnit` you just created.

nieomylnieja replied:
Changed `ReplayStatus.Unit` to `DurationUnit` and kept `ToProcessStatus` converting it back to `string` for manifest process status. Added response unmarshal coverage for `Hour`.

nieomylnieja replied:
Addressed in the current branch: ReplayStatus.Unit now uses DurationUnit.

ditrytus commented on sdk/endpoints/replay/v1/response.go:
```diff
@@ -9,21 +9,21 @@ type ReplayWithStatus struct {
-	EndTime      string `json:"endTime,omitempty"`
-	Value        int    `json:"value"`
+	Source       ReplaySource             `json:"source"`
+	Status       ReplayProcessStatus      `json:"status"`
```
This is the response where ingest will not send the values that are in `ReplayProcessStatus` enum, but the fine grained statuses that should be private. I would either change it back to `string` or update ingest.

nieomylnieja replied:
I'll update the internal code path then

nieomylnieja replied:
Kept the SDK enum on the coarse public status set (unknown, queued, in progress, completed, failed, canceled) and handled the producer-side mapping separately. I’m intentionally not linking that work from this public SDK thread.

nieomylnieja replied:
Changed ReplayStatus.Status back to string so the single-status endpoint preserves its fine-grained, open-ended values. ReplayListItem.Status remains the typed ReplayListStatus because the list endpoint returns the separate coarse status contract; the response test now covers fetching_historical_data explicitly.

nieomylnieja committed:
fix: type replay status unit

nieomylnieja committed:
feat: add replay validation and lifecycle handling

Validate replay requests before API calls, expose detailed status values, and expand replay tests to cover deletion and cancellation workflows.

nieomylnieja committed:
feat: generate replay enums and simplify validation

Replace hand-written replay enum definitions and validators with generated enums, and use enum values directly in request validation. Simplify duration conversion to return `time.Duration` without an error.

nieomylnieja committed:
fix: remove replay lifecycle test races

nieomylnieja committed:
Merge branch 'main' into add-replay-endpoints

nieomylnieja committed:
test: validate replay status values

Update replay tests for the platform's in-progress status representation and assert returned statuses belong to the supported values.

nieomylnieja committed:
fix: preserve replay creation time and validate selectors

nieomylnieja committed:
Merge remote-tracking branch 'origin/main' into add-replay-endpoints

nieomylnieja committed:
fix: align replay API contracts

nieomylnieja committed:
test: fix replay selector test names

matkaras commented on sdk/endpoints/replay/v1/request.go:
```diff
@@ -0,0 +1,427 @@
+			_, err := manifest.ParseKind(kind)
+			return err
+		}).WithErrorCode(rules.ErrorCodeOneOf)),
+	govy.For(func(r GetAvailabilityRequest) GetAvailabilityRequest { return r }).
+		Rules(rules.MutuallyExclusive(false, map[string]func(GetAvailabilityRequest) any{
+			"sloName": func(r GetAvailabilityRequest) any { return r.SLOName },
+			"dataSource": func(r GetAvailabilityRequest) any {
+				return r.DataSourceProject != "" || r.DataSource != "" || r.DataSourceKind != ""
+			},
+		})),
```
issue: The SLO details UI still calls this endpoint with both sloName and the data source query parameters. GetAvailabilityRequest.Validate() treats these selector groups as mutually exclusive, and the new test explicitly expects HTTP 400 for that request. As a result, the existing availability check for a non-composite SLO will fail. We should either update the frontend caller in this PR to send only sloName, or preserve the previous server behavior where sloName takes precedence.


example:  `https://dev-purple.nobl9.dev/api/internal/timemachine/availability?dataSourceProject=splunk-direct&dataSource=splunk-direct&dataSourceKind=Direct&sloName=aaaaasasasasdadd1&`

nieomylnieja replied:
Kept the SDK validation because SLO and data-source selectors are separate request modes.
  The companion platform change now sends only `project` and `sloName` for the SLO flow.
  The API also keeps `sloName` precedence for legacy mixed queries.
  This preserves existing callers while preventing new typed callers from sending ambiguous
  requests.
  Coverage verifies the SLO-only request, legacy precedence, and SDK rejection of mixed selectors.

matkaras commented on tests/replay_v1_test.go:
```diff
@@ -0,0 +1,330 @@
+	"github.com/nobl9/nobl9-go/tests/e2etestutils"
+)
+
+func Test_Replay_V1(t *testing.T) {
```
nitpick: it would be nice to add some comments or helper methods to make it easier to follow

nieomylnieja replied:
Split the e2e test into focused helpers for the no-existing-Replay,
  availability, cancellation, and queue-lifecycle scenarios.
  I also extracted shared status polling and cleanup.
  The top-level test now only creates fixtures and sequences those scenarios.

matkaras commented on sdk/endpoints/replay/v1/request.go:
```diff
@@ -0,0 +1,427 @@
+// TimeRange defines the earliest point from which a replay should retrieve data.
+type TimeRange struct {
+	StartDate time.Time `json:"startDate,omitzero"`
+}
```
issue: this is not time range

nieomylnieja replied:
The wire contract is `timeRange.startDate`; it has no end-date field.
Request processing time is the effective end,
so `TimeRange` still matches the JSON object rather than representing a closed interval.
I documented this semantic,
added exact `timeRange.startDate` serialization coverage,
and exercised it in the cancellation e2e scenario.
